src/uct/ib/rc/accel/rc_mlx5_ep.c

```diff
@@ -416,30 +354,161 @@ ucs_status_t uct_rc_mlx5_ep_am_zcopy(uct_ep_h tl_ep, uint8_t id, const void *hea
+                           uint64_t remote_addr, uct_rkey_t rkey,
+                           uint64_t compare_mask, uint64_t compare,
+                           uint64_t swap_mask, uint64_t swap_add, int signal)
+{
+    uct_rc_iface_t *iface  = ucs_derived_of(ep->super.super.super.iface,
+                                            uct_rc_iface_t);
+    uint32_t ib_rkey = uct_ib_resolve_atomic_rkey(rkey, ep->super.atomic_mr_offset,
+                                                  &remote_addr);
+
+    desc->super.sn = ep->tx.wq.sw_pi;
+    uct_rc_mlx5_txqp_dptr_post(iface, IBV_QPT_RC,
+                               &ep->super.txqp, &ep->tx.wq,
+                               opcode, desc + 1, length, &desc->lkey,
+                               remote_addr, ib_rkey,
+                               compare_mask, compare, swap_mask, swap_add,
+                               NULL, NULL, 0, signal, 0, INT_MAX, NULL);
+
+    UCT_TL_EP_STAT_ATOMIC(&ep->super.super);
+    uct_rc_txqp_add_send_op(&ep->super.txqp, &desc->super);
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_rc_mlx5_ep_atomic_fop(uct_rc_mlx5_ep_t *ep, int opcode, void *result, int ext,
+                          unsigned length, uint64_t remote_addr, uct_rkey_t rkey,
+                          uint64_t compare_mask, uint64_t compare,
+                          uint64_t swap_mask, uint64_t swap_add, uct_completion_t *comp)
+{
+    uct_rc_mlx5_iface_t *iface = ucs_derived_of(ep->super.super.super.iface,
+                                                uct_rc_mlx5_iface_t);
+    uct_rc_iface_send_desc_t *desc;
+
+    UCT_RC_CHECK_RES(&iface->super, &ep->super);
+    UCT_RC_IFACE_GET_TX_ATOMIC_FETCH_DESC(&iface->super, &iface->mlx5_common.tx.atomic_desc_mp,
+                                          desc, uct_rc_iface_atomic_handler(&iface->super, ext,
+                                                                            length),
+                                          result, comp);
+    uct_rc_mlx5_ep_atomic_post(ep, opcode, desc, length, remote_addr, rkey,
+                               compare_mask, compare, swap_mask, swap_add,
+                               MLX5_WQE_CTRL_CQ_UPDATE);
+    return UCS_INPROGRESS;
+}
+
+static ucs_status_t UCS_F_ALWAYS_INLINE
+uct_rc_mlx5_ep_atomic_op_post(uct_ep_h tl_ep, unsigned opcode, unsigned size,
+                              uint64_t value, uint64_t remote_addr, uct_rkey_t rkey)
+{
+    uct_rc_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_rc_mlx5_iface_t);
+    uct_rc_mlx5_ep_t *ep       = ucs_derived_of(tl_ep, uct_rc_mlx5_ep_t);
+    uct_rc_iface_send_desc_t *desc;
+    int op;
+    uint64_t compare_mask;
+    uint64_t compare;
+    uint64_t swap_mask;
+    uint64_t swap;
+    int      ext; /* not used here */
+    ucs_status_t status;
+
+    UCT_RC_MLX5_CHECK_ATOMIC_OPS(opcode);
+    ucs_assert((size == sizeof(uint64_t)) || (size == sizeof(uint32_t)));
```

Comment:
I'd add this assert to the macro above as you are using it together.

Reply:
done